src/text/line.rs

```diff
@@ -586,31 +555,84 @@ impl WidgetRef for Line<'_> {
+            // there is not enough space to render the whole line, so we need to truncate it
+            // and only render the visible part. We only care about truncating the left side of the
+            // line, as the right side will be truncated by the area width
+            let skip_width = match self.alignment {
+                Some(Alignment::Center) => (line_width.saturating_sub(area.width)) / 2,
+                Some(Alignment::Right) => line_width.saturating_sub(area.width),
                 Some(Alignment::Left) | None => 0,
             };
-            x = x.saturating_add(offset);
-            self.to_owned()
+            render_spans(&self.spans, area, buf, skip_width);
         };
-        for span in &line.spans {
-            let span_width = span.width() as u16;
-            let span_area = Rect {
-                x,
-                width: span_width.min(area.right().saturating_sub(x)),
-                ..area
-            };
-            span.render(span_area, buf);
-            x = x.saturating_add(span_width);
-            if x >= area.right() {
+    }
+}
+
+/// Renders all the spans of the line that are after the `span_offset` in the line and that are
+/// visible in the given area (starting rendering at `area.x + buf_x_offset`). `buf_x_offset`
+/// probably should just be used to modify area.x instead of being a parameter, but it's here
+/// because that would have made the above code just a little uglier
+///
+/// Algorithm:
+/// 1. For each span in the line:
+///    1. If the span is completely before the `span_skip_width`, skip it and reduce
+///       `span_skip_width` by the span's width
+///    2. If the span is partially before the `span_skip_width`, render the right side of the span
+///       starting at the span offset
+///    3. If the span is completely after the `span_skip_width`, render the whole span
+/// 2. Indent `area` to account for the width of the rendered span
+/// 3. Update `span_skip_width` to 0 after the first span is rendered in full
+/// 4. Repeat
+fn render_spans(spans: &[Span], mut area: Rect, buf: &mut Buffer, mut span_skip_width: u16) {
+    for span in spans {
+        let span_width = span.width() as u16;
+        // ignore spans that are completely before the offset by decrementing `span_skip_width` by
+        // the span width in each iteration until we find a span that is partially or completely
+        // visible
+        if span_skip_width >= span_width {
+            span_skip_width -= span_width;
+            continue;
+        }
+        if span_skip_width > 0 {
+            // the first visible span is only partially visible, so truncate the start and render it
+            // starting at span_skip_width, all subsequent visible spans are rendered in full (as
+            // available space permits). We only need to truncate the start of the first visible span
+            // because the rest of the line will be truncated automatically by the area width
```

Comment:
This is explained multiple times now.

Reply:
I generally agree that good variable names are better than comments and in my earlier years I would have probably died on the hill proclaiming an opinion that "comments are unnecessary if you write good code". Here though there's lots of ways to misinterpret the meaning of each of the variables due to the inherent complexity of what we need to do. The comments in this change serve as an apology of sorts for not being able to make the algorithm perfectly clear with just variable names. The increased quantity and verbosity of comments in an area like this also highlights the amount of intent that has gone into choosing to do things a particular way, and that needs to be understood before any refactoring / changes are made to this code.

You're right that the area truncation is explained above in the render_ref function, and that is a duplicate of this comment, but given that @kdheepak asked about this in the PR, it's reasonable to assume that someone of lesser knowledge about Ratatui who wants to edit this code in the future might have the same questions (and may not look at the calling code for answers). So I think explaining this twice doesn't do too much harm here.

Reply:
Im not against the comments explaining what is going on. Just that there is duplicated stuff now and in this case the one above is more clear to me.
There shouldn’t be duplicate explanations for the same thing.

when stuff is refactored the documentations also needs to be adapted and personally I would adapt the first thing I see describing the now changed behaviour. Another spot might be completely missed. Reviews should find something like this but it might be out of view for the changes so it might be missed.

so my general approach would be to split into good variable names and function names, then comments and doc comments. The function comment should explain the general idea, the comments on code should explain specifics why things are done the way they are done.

comments on tests explain why they are relevant to be tested and shouldn’t explain the behaviour of the function so they are out of scope in order to understand the function.

yes, @kdheepak had questions, so others might be too, but I would try to prefer pointing to the relevant documentation rather than duplicating the same comment.

(will look into it on a bigger screen)

Reply:
What triggered me here was the "rest" of the line instead of the right of the line. Then I found the same comment with "right" above.